File: html/rules.js
import { BOARD_SIZE, COLORS, PIECES } from './constants.js';

export class Rules {
    static isWithinBounds(r, c) { return r >= 0 && r < BOARD_SIZE && c >= 0 && c < BOARD_SIZE; }

    static getLegalMoves(boardState, r, c, checkKingSafety = true) {
        const piece = boardState[r][c];
        if (!piece) return [];
        let moves = [];
        const type = piece.type;
        const color = piece.color;
        const forward = color === COLORS.WHITE ? -1 : 1;

        if (type === PIECES.PAWN) {
            if (Rules.isWithinBounds(r + forward, c) && !boardState[r + forward][c]) {
                moves.push({ r: r + forward, c: c });
                const startRank = color === COLORS.WHITE ? 7 : 1;
                if (r === startRank && !boardState[r + (forward * 2)][c] && !boardState[r + forward][c]) {
                    moves.push({ r: r + (forward * 2), c: c });
                }
            }
            [[forward, -1], [forward, 1]].forEach(offset => {
                const tr = r + offset[0], tc = c + offset[1];
                if (Rules.isWithinBounds(tr, tc)) {
                    const target = boardState[tr][tc];
                    if (target && target.color !== color) moves.push({ r: tr, c: tc });
                }
            });
        }
        else if ([PIECES.ROOK, PIECES.BISHOP, PIECES.QUEEN].includes(type)) {
            const directions = [];
            if (type !== PIECES.BISHOP) directions.push([0, 1], [0, -1], [1, 0], [-1, 0]);
            if (type !== PIECES.ROOK) directions.push([1, 1], [1, -1], [-1, 1], [-1, -1]);
            directions.forEach(dir => {
                for (let i = 1; i < BOARD_SIZE; i++) {
                    const tr = r + (dir[0] * i), tc = c + (dir[1] * i);
                    if (!Rules.isWithinBounds(tr, tc)) break;
                    const target = boardState[tr][tc];
                    if (!target) moves.push({ r: tr, c: tc });
                    else { if (target.color !== color) moves.push({ r: tr, c: tc }); break; }
                }
            });
        }
        else { // Knight, King, Krishna
            let offsets = (type === PIECES.KNIGHT)
                ? [[-2, -1], [-2, 1], [-1, -2], [-1, 2], [1, -2], [1, 2], [2, -1], [2, 1]]
                : [[-1, -1], [-1, 0], [-1, 1], [0, -1], [0, 1], [1, -1], [1, 0], [1, 1]];

            offsets.forEach(off => {
                const tr = r + off[0], tc = c + off[1];
                if (Rules.isWithinBounds(tr, tc)) {
                    const target = boardState[tr][tc];
                    if (type === PIECES.KRISHNA) { if (!target) moves.push({ r: tr, c: tc }); }
                    else if (!target || target.color !== color) moves.push({ r: tr, c: tc });
                }
            });
        }

        moves = moves.filter(m => {
            const t = boardState[m.r][m.c];
            return !(t && t.type === PIECES.KRISHNA);
        });

        if (checkKingSafety) {
            moves = moves.filter(m => {
                const savedTarget = boardState[m.r][m.c];
                const savedSource = boardState[r][c];
                boardState[m.r][m.c] = savedSource;
                boardState[r][c] = null;
                const inCheck = Rules.isKingInCheck(boardState, color);
                boardState[r][c] = savedSource;
                boardState[m.r][m.c] = savedTarget;
                return !inCheck;
            });
        }
        return moves;
    }

    static isKingInCheck(board, color) {
        let kR, kC;
        for (let r = 0; r < BOARD_SIZE; r++) {
            for (let c = 0; c < BOARD_SIZE; c++) {
                if (board[r][c] && board[r][c].color === color && board[r][c].type === PIECES.KING) {
                    kR = r; kC = c; break;
                }
            }
        }
        if (kR === undefined) return true;
        const enemyColor = color === COLORS.WHITE ? COLORS.BLACK : COLORS.WHITE;

        // Check Rays
        const dirs = [[0, 1], [0, -1], [1, 0], [-1, 0], [1, 1], [1, -1], [-1, 1], [-1, -1]];
        for (let dir of dirs) {
            for (let i = 1; i < BOARD_SIZE; i++) {
                const tr = kR + (dir[0] * i), tc = kC + (dir[1] * i);
                if (!Rules.isWithinBounds(tr, tc)) break;
                const p = board[tr][tc];
                if (p) {
                    if (p.color === enemyColor) {
                        const t = p.type;
                        if (t === PIECES.KRISHNA) break;
                        const isOrtho = (dir[0] === 0 || dir[1] === 0);
                        if (isOrtho && (t === PIECES.ROOK || t === PIECES.QUEEN)) return true;
                        if (!isOrtho && (t === PIECES.BISHOP || t === PIECES.QUEEN)) return true;
                        if (i === 1 && t === PIECES.KING) return true;
                        if (i === 1 && !isOrtho && t === PIECES.PAWN) {
                            if (enemyColor === COLORS.WHITE && dir[0] === 1) return true;
                            if (enemyColor === COLORS.BLACK && dir[0] === -1) return true;
                        }
                    }
                    break;
                }
            }
        }
        // Check Knights
        const knightOffsets = [[-2, -1], [-2, 1], [-1, -2], [-1, 2], [1, -2], [1, 2], [2, -1], [2, 1]];
        for (let off of knightOffsets) {
            const tr = kR + off[0], tc = kC + off[1];
            if (Rules.isWithinBounds(tr, tc)) {
                const p = board[tr][tc];
                if (p && p.color === enemyColor && p.type === PIECES.KNIGHT) return true;
            }
        }
        return false;
    }
}

File: crates/shatrunz-core/src/rules.rs
//! Faithful port of `frontend/rules.js`. Iteration order (row-major board
//! scan, per-piece offset order, direction order) is preserved exactly —
//! it feeds search move ordering later (see docs/RUST_PORT.md "Rules /
//! movement": "exact JS move-generation order replicated").

use crate::board::{in_bounds, Board, BOARD_SIZE};
use crate::moves::{CastlingSide, Move, MoveTarget, Square};
use crate::piece::{Color, PieceKind};

/// Mirrors `Rules.getAllLegalMoves` (rules.js:6-19). Takes `&Board`
/// (read-only from the caller's point of view) — internally uses one
/// mutable scratch copy for the make/unmake king-safety check, same as
/// JS mutating `this.board` in place and always reverting.
pub fn get_all_legal_moves(board: &Board, color: Color) -> Vec<Move> {
    let mut scratch = board.clone();
    let mut moves = Vec::new();
    for r in 0..BOARD_SIZE {
        for c in 0..BOARD_SIZE {
            match scratch.get(r, c) {
                Some(p) if p.color == color => {}
                _ => continue,
            }
            for to in get_legal_moves_mut(&mut scratch, r, c, true) {
                moves.push(Move {
                    from: Square { r, c },
                    to,
                });
            }
        }
    }
    moves
}

/// Mirrors `Rules.getLegalMoves` (rules.js:21-134). Takes `&Board`; see
/// `get_all_legal_moves` for why this is a cloning wrapper.
pub fn get_legal_moves(
    board: &Board,
    r: usize,
    c: usize,
    check_king_safety: bool,
) -> Vec<MoveTarget> {
    let mut scratch = board.clone();
    get_legal_moves_mut(&mut scratch, r, c, check_king_safety)
}

fn get_legal_moves_mut(
    board: &mut Board,
    r: usize,
    c: usize,
    check_king_safety: bool,
) -> Vec<MoveTarget> {
    let piece = match board.get(r, c) {
        Some(p) => p,
        None => return Vec::new(),
    };
    let kind = piece.kind;
    let color = piece.color;
    let forward: isize = if color == Color::White { -1 } else { 1 };
    let ir = r as isize;
    let ic = c as isize;

    let mut moves: Vec<MoveTarget> = Vec::new();

    match kind {
        PieceKind::Pawn => {
            if in_bounds(ir + forward, ic) && board.get((ir + forward) as usize, c).is_none() {
                moves.push(MoveTarget::plain((ir + forward) as usize, c));
                let start_rank: isize = if color == Color::White { 7 } else { 1 };
                if ir == start_rank
                    && board
                        .get((ir + forward * 2) as usize, c)
                        .is_none()
                    && board.get((ir + forward) as usize, c).is_none()
                {
                    moves.push(MoveTarget::plain((ir + forward * 2) as usize, c));
                }
            }
            for (dr, dc) in [(forward, -1isize), (forward, 1isize)] {
                let tr = ir + dr;
                let tc = ic + dc;
                if in_bounds(tr, tc) {
                    if let Some(target) = board.get(tr as usize, tc as usize) {
                        if target.color != color {
                            moves.push(MoveTarget::plain(tr as usize, tc as usize));
                        }
                    }
                }
            }
        }
        PieceKind::Rook | PieceKind::Bishop | PieceKind::Queen => {
            let mut directions: Vec<(isize, isize)> = Vec::new();
            if kind != PieceKind::Bishop {
                directions.extend([(0, 1), (0, -1), (1, 0), (-1, 0)]);
            }
            if kind != PieceKind::Rook {
                directions.extend([(1, 1), (1, -1), (-1, 1), (-1, -1)]);
            }
            for (dr, dc) in directions {
                for i in 1..BOARD_SIZE as isize {
                    let tr = ir + dr * i;
                    let tc = ic + dc * i;
                    if !in_bounds(tr, tc) {
                        break;
                    }
                    match board.get(tr as usize, tc as usize) {
                        None => moves.push(MoveTarget::plain(tr as usize, tc as usize)),
                        Some(target) => {
                            if target.color != color {
                                moves.push(MoveTarget::plain(tr as usize, tc as usize));
                            }
                            break;
                        }
                    }
                }
            }
        }
        PieceKind::Knight | PieceKind::King | PieceKind::Krishna => {
            let offsets: &[(isize, isize)] = if kind == PieceKind::Knight {
                &[
                    (-2, -1),
                    (-2, 1),
                    (-1, -2),
                    (-1, 2),
                    (1, -2),
                    (1, 2),
                    (2, -1),
                    (2, 1),
                ]
            } else {
                &[
                    (-1, -1),
                    (-1, 0),
                    (-1, 1),
                    (0, -1),
                    (0, 1),
                    (1, -1),
                    (1, 0),
                    (1, 1),
                ]
            };
            for (dr, dc) in offsets {
                let tr = ir + dr;
                let tc = ic + dc;
                if in_bounds(tr, tc) {
                    let target = board.get(tr as usize, tc as usize);
                    if kind == PieceKind::Krishna {
                        if target.is_none() {
                            moves.push(MoveTarget::plain(tr as usize, tc as usize));
                        }
                    } else if target.is_none() || target.unwrap().color != color {
                        moves.push(MoveTarget::plain(tr as usize, tc as usize));
                    }
                }
            }
        }
    }

    // Krishna can never be captured, by any piece type (rules.js:74-77).
    moves.retain(|m| !matches!(board.get(m.r, m.c), Some(t) if t.kind == PieceKind::Krishna));

    if check_king_safety {
        moves.retain(|m| {
            let saved_target = board.get(m.r, m.c);
            let saved_source = board.get(r, c);
            board.set(m.r, m.c, saved_source);
            board.set(r, c, None);
            let in_check = is_king_in_check(board, color);
            board.set(r, c, saved_source);
            board.set(m.r, m.c, saved_target);
            !in_check
        });

        // Castling: king steps two squares; JS is position-only (no
        // rights tracking) — see docs/RUST_PORT.md "Rules / movement".
        if kind == PieceKind::King && !is_king_in_check(board, color) {
            let king_rank: usize = if color == Color::White { 8 } else { 0 };
            if r == king_rank && c == 4 {
                // Kingside: F(5), G(6) empty; rook at I(8).
                if board.get(king_rank, 5).is_none() && board.get(king_rank, 6).is_none() {
                    if let Some(rook) = board.get(king_rank, 8) {
                        if rook.kind == PieceKind::Rook
                            && rook.color == color
                            && !is_square_attacked(board, king_rank, 5, color)
                            && !is_square_attacked(board, king_rank, 6, color)
                        {
                            moves.push(MoveTarget::castling(king_rank, 6, CastlingSide::King));
                        }
                    }
                }
                // Queenside: D(3), C(2), B(1) empty; rook at A(0).
                if board.get(king_rank, 3).is_none()
                    && board.get(king_rank, 2).is_none()
                    && board.get(king_rank, 1).is_none()
                {
                    if let Some(rook) = board.get(king_rank, 0) {
                        if rook.kind == PieceKind::Rook
                            && rook.color == color
                            && !is_square_attacked(board, king_rank, 3, color)
                            && !is_square_attacked(board, king_rank, 2, color)
                        {
                            moves.push(MoveTarget::castling(king_rank, 2, CastlingSide::Queen));
                        }
                    }
                }
            }
        }
    }

    moves
}

const KNIGHT_OFFSETS: [(isize, isize); 8] = [
    (-2, -1),
    (-2, 1),
    (-1, -2),
    (-1, 2),
    (1, -2),
    (1, 2),
    (2, -1),
    (2, 1),
];

const RAY_DIRECTIONS: [(isize, isize); 8] = [
    (0, 1),
    (0, -1),
    (1, 0),
    (-1, 0),
    (1, 1),
    (1, -1),
    (-1, 1),
    (-1, -1),
];

/// Mirrors `Rules.isSquareAttacked` (rules.js:136-195): is `(r, c)`
/// attacked by any `ally_color`-opposing piece? Krishna blocks rays but
/// never itself attacks (it has no capture moves, so it's excluded from
/// both the ray and knight scans by construction).
pub fn is_square_attacked(board: &Board, r: usize, c: usize, ally_color: Color) -> bool {
    let enemy = ally_color.opposite();
    let ir = r as isize;
    let ic = c as isize;

    for (dr, dc) in KNIGHT_OFFSETS {
        let tr = ir + dr;
        let tc = ic + dc;
        if in_bounds(tr, tc) {
            if let Some(p) = board.get(tr as usize, tc as usize) {
                if p.color == enemy && p.kind == PieceKind::Knight {
                    return true;
                }
            }
        }
    }

    for (dr, dc) in RAY_DIRECTIONS {
        for i in 1..BOARD_SIZE as isize {
            let tr = ir + dr * i;
            let tc = ic + dc * i;
            if !in_bounds(tr, tc) {
                break;
            }
            if let Some(p) = board.get(tr as usize, tc as usize) {
                if p.color == enemy {
                    if p.kind == PieceKind::Krishna {
                        break;
                    }
                    let is_ortho = dr == 0 || dc == 0;
                    if is_ortho && (p.kind == PieceKind::Rook || p.kind == PieceKind::Queen) {
                        return true;
                    }
                    if !is_ortho && (p.kind == PieceKind::Bishop || p.kind == PieceKind::Queen) {
                        return true;
                    }
                    if i == 1 && p.kind == PieceKind::King {
                        return true;
                    }
                    if i == 1 && !is_ortho && p.kind == PieceKind::Pawn {
                        if enemy == Color::White && dr == 1 {
                            return true;
                        }
                        if enemy == Color::Black && dr == -1 {
                            return true;
                        }
                    }
                }
                break;
            }
        }
    }

    false
}

/// Mirrors `Rules.isKingInCheck` (rules.js:197-243). Returns true if the
/// king is absent from the board (rules.js:206) — a deliberate quirk kept
/// for parity, not a bug.
pub fn is_king_in_check(board: &Board, color: Color) -> bool {
    let mut king_sq: Option<(usize, usize)> = None;
    'search: for r in 0..BOARD_SIZE {
        for c in 0..BOARD_SIZE {
            if let Some(p) = board.get(r, c) {
                if p.color == color && p.kind == PieceKind::King {
                    king_sq = Some((r, c));
                    break 'search;
                }
            }
        }
    }
    match king_sq {
        None => true,
        Some((r, c)) => is_square_attacked(board, r, c, color),
    }
}
